torch/testing/_asserts.py

```diff
@@ -102,6 +102,36 @@ def wrapper(actual: Tensor, expected: Tensor, **kwargs: Any) -> Optional[Excepti
     return wrapper
 
 
+def _check_sparse_indices_and_values_individually(
+    check_tensors: Callable[..., Optional[Exception]]
+) -> Callable[..., Optional[Exception]]:
+    """Decorates dense tensor check functions to handle indices and values of sparse tensors individually.
+
+    If the inputs are not sparse, this decorator is a no-op.
+
+    Args:
+        check_tensors (Callable[[Tensor, Tensor], Optional[Exception]]): Tensor check function for dense
+        tensors.
+    """
+    @functools.wraps(check_tensors)
+    def wrapper(actual: Tensor, expected: Tensor, **kwargs: Any) -> Optional[Exception]:
+        if not actual.is_sparse:
+            return check_tensors(actual, expected, **kwargs)
+
+        actual = actual.coalesce()
+        expected = expected.coalesce()
```

Comment:
@cpuhrsch @pearu, should sparse tensors always be coalesced before they're compared for "closeness?"

Reply:
@mruberry  technically, calling `coalesce` is required when one uses `values()` and `indices()` for checking "closeness".

On the other hand, one can have two unequal uncoalesced tensors ("closeness" is check using `_values()` and `_indices()`) that after coalescing become equal.

So, whether to coalesce or not for "closeness" checks, depends on what is being tested. For instance, different algorithms may produce unequal uncoalesced tensors while the coalesced tensors end up being equal. On the other hand, if the process of producing uncoalesced tensors needs to be tested, then one should not coalesce for checking closeness. Although, I cannot think of an application to this atm..

Reply:
I think we can solve this with an `check_coalesced: bool = False` keyword argument as proposed in https://github.com/pytorch/pytorch/pull/58844#discussion_r643091290. By default this would call `coalesce()` on both tensors, but could be disabled, if we later hit a case where we need the difference.